addons/sale/models/sale_order.py

```diff
@@ -1709,6 +1795,104 @@ def _get_default_payment_link_values(self):
         }
 
+    # EDI #
+
+    def create_document_from_attachment(self, attachment_ids):
+        """ Create the sale orders from given attachment_ids and redirect newly create order view.
+
+        :param list attachment_ids: List of attachments process.
+        :return: An action redirecting to related sale order view.
+        :rtype: dict
+        """
+        orders = self._create_order_from_attachment(attachment_ids)
+        action_vals = {
+            'name': _("Generated Orders"),
+            'domain': [('id', 'in', orders.ids)],
+            'res_model': 'sale.order',
+            'type': 'ir.actions.act_window',
+            'context': self._context
+        }
+        if len(orders) == 1:
+            action_vals.update({
+                'views': [[False, 'form']],
+                'res_id': orders.id,
+            })
+        else:
+            action_vals.update({
+                'views': [[False, 'list'], [False, 'kanban'], [False, 'form']],
+            })
+
+        return action_vals
+
+    @api.model
+    def _create_order_from_attachment(self, attachment_ids):
+        """ Create the sale orders from given attachment_ids and fill data by extracting detail
+        from attachments and return generated orders.
+
+        :param list attachment_ids: List of attachments process.
+        :return: Recordset of order.
+        """
+        attachments = self.env['ir.attachment'].browse(attachment_ids)
+        if not attachments:
+            raise UserError(_("No attachment was provided"))
+
+        orders = self
+        for attachment in attachments:
+            order = self.create({
+                'partner_id': self.env.user.partner_id.id,
+            })
+            order._extend_with_attachments(attachment)
+            orders |= order
+            order.message_post(attachment_ids=attachment.ids)
+            attachment.write({'res_model': self._name, 'res_id': order.id})
+
+        return orders
+
+    def _extend_with_attachments(self, attachment):
+        """ Main entry point to extend/enhance order with attachment.
+
+        :param attachment: A recordset of ir.attachment.
+        :returns: None
+        """
+        self.ensure_one()
+
+        def close_file(file_data):
+            if file_data.get('on_close'):
+                file_data['on_close']()
+
+        file_data = attachment._unwrap_edi_attachments()[0]
+        decoder = self._get_order_edi_decoder(file_data)
+        if decoder:
+            try:
+                with self.env.cr.savepoint():
+                    decoder(self, file_data)
+            except RedirectWarning:
+                raise
+            except Exception:
+                message = _(
+                    "Error importing attachment '%(file_name)s' as order (decoder=%(decoder)s)",
+                    file_name=file_data['filename'],
+                    decoder=decoder.__name__,
+                )
+                self.sudo().message_post(body=message)
```

Comment:
why the sudo here?

Reply:
Because message should be logged from odoobot :thinking:

Reply:
That's not `sudo` for that, that's `with_user(SUPERUSER_ID)` :eyes:

Reply:
Yes I know :sweat_smile: 
But I also think may be in future we might create SO from email like invoice or actually use peppol end points on customer where sending doc of Purchase order file fetched and create Sale Order by system (I don't how actually peppol end points works :see_no_evil: ) :thinking: 
But I think I should use `with_user(SUPERUSER_ID)` for now :see_no_evil: